### PR TITLE
Enhancements to M43 pins debugging

### PR DESCRIPTION
**Pull request by thinkyhead**
Cleanup and optimization of #5112
- Add endstop monitoring option with `E` parameter
- Report PWM status of pins
- Give an extended pin report when not watching with `W`

### CODE REVIEW

File: Marlin/pinsDebug.h [hunk 1]
```diff
@@ -21,8 +21,9 @@
  */
 
 // How many DIO pins are defined?
-#if defined(DIO85_PIN)
-  #define DIO_COUNT 86
+#ifdef DIO85_PIN
+//  #define DIO_COUNT 86
+  #define DIO_COUNT 70  // digitalRead and other Arduino IDE routines only know about pins 0 through 69
```
Comment (Blue-Marlin) [inline Marlin/pinsDebug.h]: The `pins_arduino.h` files provide this info in `NUM_DIGITAL_PINS` and `NUM_ANALOG_INPUTS`

File: Marlin/temperature.cpp [hunk 2]
```diff
@@ -1848,4 +1932,11 @@ void Temperature::isr() {
       }
     }
   #endif //BABYSTEPPING
+  #if ENABLED(PINS_DEBUGGING)
+    // run the endstop monitor at 15Hz
+    static uint8_t endstop_monitor_count = 16;  // offset this check from the others
+    endstop_monitor_count += _BV(1);  //  15 Hz
+    endstop_monitor_count &= 0x7F;
+    if (endstop_monitor_count == 0) endstop_monitor();  // report changes in endstop status
+  #endif
```
Comment (Blue-Marlin) [inline Marlin/temperature.cpp]: `endstop_monitor_flag` could be checked here and would save the call of `endstop_monitor()` if not active.

Comment (thinkyhead) [reply Marlin/temperature.cpp]: Thanks for the tip. I totally forgot to review that part of the code.

**Comment by Bob-the-Kuhn** [comment]
Much cleaner code

I've got a couple of bug fixes & a couple of minor changes.

All the changes are included in the attached file.
[pinsDebug .zip](https://github.com/MarlinFirmware/Marlin/files/562502/pinsDebug.zip)

I couldn't get the  **#define PINSET(P) (defined(P) && P > -1)**  from the previous review to work on a pin that existed.  Really strange since it uses the same method as the PIN_EXISTS macro.

# 

When determining if a pin is connected to a PWM, the TCCRxA register is always used even if looking at TIMERxC

Lines 677-680 need to change

```
#define PWM_CASE(N) \
  case TIMER##N: \
    if (TCCR##N & (_BV(COM## N ##1) | _BV(COM## N ##0))) { \
      PWM_PRINT(OCR##N); \
```

The solution I came up with is a two part change:
Lines 677-680  now read

```
#define PWM_CASE(N,Z) \
  case TIMER##N##Z: \
    if (TCCR##N ##A & (_BV(COM## N ## Z ##1) | _BV(COM## N ## Z ##0))) { \
      PWM_PRINT(OCR##N ## Z); \
```

which means 694-724 change from
`PWM_CASE(0A);`
to
`PWM_CASE(0,A);`

# 

In the WGM_MAKE macros the TEST function only returns 1 or 0 so the following pieces in lines 734-735 need to change from

```
TEST(TCCR##N##B, WGM##N##2) >> 1
TEST(TCCR##N##B, WGM##N##3) >> 1
```

 to either 

```
TEST(TCCR##N##B, WGM##N##2) << 2
TEST(TCCR##N##B, WGM##N##3) << 3
```

or

```
TCCR##N##B & _BV(WGM##N##2)) >> 1
TCCR##N##B & _BV(WGM##N##3)) >> 1
```

I prefer the latter because it's one less operation to perform and it's consistent with how the macros treat the other register.

# 

A couple of minor changes:
- buffer[30] changed to buffer[40] per review in previous PR
- added name length to comment on line 75
  `// Pin list updated from 7 OCT RCBugfix branch   - max length of pin name is 24`

**Comment by thinkyhead** [comment]
> Really strange since it uses the same method as the `PIN_EXISTS` macro.

Almost. When `PIN_EXISTS` tries to evaluate an undefined symbol the result is:

``` cpp
defined(_PIN) && _PIN >= 0
```

…but with the proposed `PINSET` it ends up being:

``` cpp
defined() && >= 0
```

…which is nonsense. I hit this before, but I forgot.

**Comment by thinkyhead** [comment]
Ah, you're right about my mis-use of `TEST`. I forgot it simply evaluates as 1 or 0, and not as the bit.

**Comment by thinkyhead** [comment]
> buffer[30] changed to buffer[40] per review in previous PR

I think 30 is fine. I forgot that justified strings already include the string length.

**Pull request merged**